[WebGPU] webgpu:web_platform,copyToTexture,ImageData:* is failing

ruthvikkonda opened the pull request:
#### a05bb82b74278830f51aea6945abf5e9672a9ee7
<pre>
[WebGPU] webgpu:web_platform,copyToTexture,ImageData:* is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=298141">https://bugs.webkit.org/show_bug.cgi?id=298141</a>
<a href="https://rdar.apple.com/159491345">rdar://159491345</a>

Reviewed by Mike Wyrzykowski.

Currently, when copying external image bytes to the destination texture,
we use the same behavior for the Bgra8unorm and Bgra8unormSRGB texture formats
that we do for the Rgba8unorm formats. This is incorrect. We should be
performing a conversion here for the appropriate channels. This fix
does that.

Additionally, getImageBytesFromImageBuffer() should never do pixel format conversion. It should
just provide the RGBA8 bytes to copyToDestinationFormat(), and then
copyToDestinationFormat() should be doing all pixel format conversions.

These changes also fixed various subtest cases in image_file.html and image.html
so expected.txt files were updated. Lastly, we update mac-wk2
TestExpectations to run all web_platform/copyToTexture tests so we can catch
subtest regressions.

* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/ImageData-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/image-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/image_file-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromImageBuffer):
(WebCore::imageBytesForSource):
(WebCore::copyToDestinationFormat):

Canonical link: <a href="https://commits.webkit.org/299914@main">https://commits.webkit.org/299914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a4a4a70f958801121d19062f162ec08ca45c257

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120562 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126943 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72637 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/25f7818f-51ad-47ee-aeb0-d68741718e2e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48833 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91561 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60828 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aaa13d6e-b13e-442c-8afd-d68f94fd22dc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108068 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72111 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b46cebe1-12fe-42bb-8b2e-74c021c34582) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31722 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26173 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70556 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26355 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129823 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47483 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36039 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100178 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47850 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104249 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100020 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25412 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45463 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23491 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44131 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47345 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53050 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46813 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50160 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48500 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/02beb2e0250d69f02b474d1fc6c63b33e547d993. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50350

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a32f5895adfa103075fe51947c4c117f9776f06e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50350

mwyrzykowski reviewed (changes requested):
Actually I think this regresses a test.

Without this change:
https://gpuweb.github.io/cts/standalone/?q=webgpu:web_platform,copyToTexture,ImageBitmap:from_ImageData:alpha=%22none%22;orientation=%22none%22;colorSpaceConversion=%22none%22;srcFlipYInCopy=true;dstFormat=%22bgra8unorm%22;*

passes.

With this change, it fails.

I'm not sure why the test isn't running as part of EWS as I observed it running:

./Tools/Scripts/run-webkit-tests --release --no-build --reset-results

locally

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff6f111bea5a9c9b237af7ff2266a058a54caee9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50350

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dc6812a16e96d0c5a5dcf2bbe82c550dac452fcd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50350

ruthvikkonda committed:
[WebGPU] webgpu:web_platform,copyToTexture,ImageData:* is failing
https://bugs.webkit.org/show_bug.cgi?id=298141
rdar://159491345

Reviewed by Mike Wyrzykowski.

Currently, when copying external image bytes to the destination texture,
we use the same behavior for the Bgra8unorm and Bgra8unormSRGB texture formats
that we do for the Rgba8unorm formats. This is incorrect. We should be
performing a conversion here for the appropriate channels. This fix
does that.

Additionally, getImageBytesFromImageBuffer() should never do pixel format conversion. It should
just provide the RGBA8 bytes to copyToDestinationFormat(), and then
copyToDestinationFormat() should be doing all pixel format conversions.

These changes also fixed various subtest cases in image_file.html and image.html
so expected.txt files were updated. Lastly, we update mac-wk2
TestExpectations to run all web_platform/copyToTexture tests so we can catch
subtest regressions.

* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/ImageData-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/image-expected.txt:
* LayoutTests/http/tests/webgpu/webgpu/web_platform/copyToTexture/image_file-expected.txt:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::getImageBytesFromImageBuffer):
(WebCore::imageBytesForSource):
(WebCore::copyToDestinationFormat):

Canonical link: https://commits.webkit.org/299914@main
